service/src/test/java/bio/terra/externalcreds/services/FenceProviderServiceTest.java

```diff
@@ -66,11 +87,113 @@ void testGetLinkedFenceAccountKey() {
     when(fenceAccountKeyService.getFenceAccountKey(linkedAccount))
         .thenReturn(Optional.of(fenceAccountKey));
 
-    var actualFenceAccountKey = fenceProviderService.getFenceAccountKey(userId, provider);
+    var actualFenceAccountKey = fenceProviderService.getFenceAccountKey(linkedAccount);
 
     assertPresent(actualFenceAccountKey);
   }
 
+  @Test
+  void testCreateLink() {
+    var linkedAccountId = random.nextInt();
+    var linkedAccount = TestUtils.createRandomLinkedAccount(Provider.FENCE);
+    when(linkedAccountService.upsertLinkedAccount(any()))
+        .thenReturn(linkedAccount.withId(linkedAccountId));
+    var auditLogEventBuilder = new AuditLogEvent.Builder().userId(linkedAccount.getUserId());
+    var oauth2State =
+        new OAuth2State.Builder()
+            .provider(Provider.FENCE)
+            .random("abcde")
+            .redirectUri("http://localhost:8080/oauth2/callback")
+            .build();
+    var encodedState = oauth2State.encode(objectMapper);
+
+    var providerClient =
+        ClientRegistration.withRegistrationId(linkedAccount.getProvider().toString())
+            .authorizationGrantType(AuthorizationGrantType.REFRESH_TOKEN)
+            .build();
+    when(providerOAuthClientCache.getProviderClient(linkedAccount.getProvider()))
+        .thenReturn(providerClient);
+    doNothing()
+        .when(linkedAccountService)
+        .validateAndDeleteOAuth2State(linkedAccount.getUserId(), oauth2State);
+
+    var provider = TestUtils.createRandomProvider().setScopes(Set.of("scope"));
+    when(externalCredsConfig.getProviderProperties(linkedAccount.getProvider()))
+        .thenReturn(provider);
+    var accessTokenResponse =
+        OAuth2AccessTokenResponse.withToken("token")
+            .refreshToken("refresh")
+            .tokenType(OAuth2AccessToken.TokenType.BEARER)
+            .build();
+    when(oAuth2Service.authorizationCodeExchange(
+            providerClient,
+            "code",
+            "http://localhost:8080/oauth2/callback",
+            Set.copyOf(provider.getScopes()),
+            encodedState,
+            Map.of()))
+        .thenReturn(accessTokenResponse);
+
+    when(oAuth2Service.getUserInfo(providerClient, accessTokenResponse.getAccessToken()))
+        .thenReturn(
+            new DefaultOAuth2User(
+                null,
+                Map.of("foo", "bar", provider.getExternalIdClaim(), "barName"),
+                provider.getExternalIdClaim()));
+
+    fenceProviderService.createLink(
```

Comment:
Can we assert that the created linked account is what we expect?

Reply:
Done!